PopPUNK/lineage_clustering.py

```diff
@@ -0,0 +1,680 @@
+        if qlist is not None:
+            lFile.write(',Status')
+        lFile.write('\n')
+
+        # print lines for each isolate
+        for isolate in lineage_clustering[R].keys():
+            lFile.write(isolate)
+            for n,R in enumerate(rank_list):
+                lFile.write(',' + str(lineage_clustering[R][isolate]))
+                lineage_string = str(lineage_clustering[R][isolate])
+                # include information on lineage clustering
+                combined[titles_list[n]][isolate] = lineage_string
+                if lineage_clustering[R][isolate] != previous_lineage_clustering[R][isolate] and previous_lineage_clustering[R][isolate] != null_cluster_value:
+                    lineage_string = str(previous_lineage_clustering[R][isolate]) + ':' + lineage_string
+                if isolate in combined['Overall_lineage'].keys():
+                    combined['Overall_lineage'][isolate] = combined['Overall_lineage'][isolate] + '-' + lineage_string
+                else:
+                    combined['Overall_lineage'][isolate] = lineage_string
+            lFile.write(',' + combined['Overall_lineage'][isolate])
+            if qlist is not None:
+                if isolate in qlist:
+                    lFile.write(',Added')
+                else:
+                    lFile.write(',Existing')
+            lFile.write('\n')
+
+    return combined
+
+def run_clustering_for_R(R, null_cluster_value = None, qlist = None, existing_scheme = False, distances_length = None, distances_type = None):
+    """ Clusters isolates into lineages based on their
+    relative distances using a single R to enable
+    parallelisation.
+
+    Args:
+        R (int)
+            Integer specifying the maximum rank of neighbour used
+            for clustering. Should be changed to int list for hierarchical
+            clustering.
+        null_cluster_value (int)
+            Used to denote as-yet-unclustered isolates.
+        qlist (list)
+            List of query sequences being added to an existing clustering.
+            Should be included within rlist.
+        use_existing (bool)
+            Whether to extend a previously generated analysis or not.
+            
+    Returns:
+        lineage_clustering (dict)
+            Assignment of each isolate to a cluster.
+        lineage_seed (dict)
+            Seed isolate used to initiate each cluster.
+        neighbours (nested dict)
+            Neighbour relationships between isolates for R.
+    """
+    
+    # load shared memory objects
+    existing_distance_shm = shared_memory.SharedMemory(name = 'shm_distances')
+    distances = np.ndarray(distances_length, dtype = distances_type, buffer = existing_distance_shm.buf)
+    existing_rank_shm = shared_memory.SharedMemory(name = 'shm_distance_ranks')
+    distance_ranks = np.ndarray(distances_length, dtype = 'int64', buffer = existing_rank_shm.buf)
+    isolate_list = shared_memory.ShareableList(name = 'shm_isolate_list')
```

Comment:
Not sure how I feel about sharable list being essentially a global here. Use with Pool might be better

Reply:
Added in da5b6df826fe1cafe622094f82cb8d3677354661.